subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/IndexedNormalizedFileSnapshot.java

```diff
@@ -38,4 +38,32 @@ public String getAbsolutePath() {
     public int getIndex() {
         return index;
     }
+
+    @Override
+    protected boolean hasSameNormalizedPathAs(AbstractNormalizedFileSnapshot that) {
+        if (that instanceof IndexedNormalizedFileSnapshot) {
+            IndexedNormalizedFileSnapshot other = (IndexedNormalizedFileSnapshot) that;
+            int myLength = absolutePath.length() - index;
+            int otherLength = other.absolutePath.length() - other.index;
+            if (myLength != otherLength) {
+                return false;
+            }
+            for (int i = myLength - 1; i >= 0; i--) {
+                if (absolutePath.charAt(index + i) != other.absolutePath.charAt(other.index + i)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    protected int hashNormalizedPath() {
+        int h = 0;
+        for (int i = index; i < absolutePath.length(); i++) {
```

Comment:
:o: Why not directly use `absolutePath.hashCode()` here? Why limiting the hashcode to the relative path?

Reply:
Because by definition snapshots are compared by comparing relative path and content hash.

Reply:
I get it, but there's a performance impact of using a custom hash, especially for strings. The hash code will only ensure a good distribution. If the amount of chars you use is lower, there's a higher risk of collision.

Reply:
This whole class is just a memory optimization - what we really want is a relative path, but we store it as a combination of interned absolute path and an index. But for the purpose of equality/hashCode we want to treat it represented a relative path.

Reply:
We want ("foobar", 3) and ("bar", 0) to be considered equal, that's why we can't use the hash of the absolute path.